File: frontend/src/types/application.ts
// =========================================================
// Application Status
// =========================================================

export type ApplicationStatus =
  | "saved"
  | "applied"
  | "screening"
  | "interview"
  | "offer"
  | "accepted"
  | "rejected"
  | "withdrawn";


// =========================================================
// Application
// =========================================================

export interface Application {
  id: string;

  user_id: string;

  // -------------------------------------------------------
  // Job information
  // -------------------------------------------------------

  job_id: string | null;

  job_title: string;

  company: string;

  location: string | null;

  apply_url: string | null;

  // -------------------------------------------------------
  // Application state
  // -------------------------------------------------------

  status: ApplicationStatus;

  // -------------------------------------------------------
  // User notes
  // -------------------------------------------------------

  notes: string | null;

  // -------------------------------------------------------
  // Dates
  // -------------------------------------------------------

  applied_at: string | null;

  created_at: string;

  updated_at: string;
}

File: backend/Applications/schemas/application.py
"""
Application Schemas

Defines the domain models used by the application tracking
and Kanban workflow.
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


# =========================================================
# APPLICATION STATUS
# =========================================================

class ApplicationStatus(str, Enum):
    """
    Current stage of a job application.
    """

    SAVED = "saved"

    APPLIED = "applied"

    SCREENING = "screening"

    INTERVIEW = "interview"

    OFFER = "offer"

    ACCEPTED = "accepted"

    REJECTED = "rejected"

    WITHDRAWN = "withdrawn"


# =========================================================
# APPLICATION
# =========================================================

class Application(BaseModel):
    """
    Represents one job application belonging to a user.
    """

    id: str

    user_id: str

    # -----------------------------------------------------
    # Job information
    # -----------------------------------------------------

    job_id: Optional[str] = None

    job_title: str

    company: str

    location: Optional[str] = None

    apply_url: Optional[str] = None

    # -----------------------------------------------------
    # Application state
    # -----------------------------------------------------

    status: ApplicationStatus = (
        ApplicationStatus.SAVED
    )

    # -----------------------------------------------------
    # User notes
    # -----------------------------------------------------

    notes: Optional[str] = None

    # -----------------------------------------------------
    # Dates
    # -----------------------------------------------------

    applied_at: Optional[datetime] = None

    created_at: datetime = Field(
        default_factory=datetime.utcnow
    )

    updated_at: datetime = Field(
        default_factory=datetime.utcnow
    )